dio/lib/src/failed_request_client_adapter.dart

```diff
@@ -0,0 +1,164 @@
+  /// - In an mobile or desktop application this can be requests which failed
+  ///   because the connection was interrupted.
+  final bool captureFailedRequests;
+
+  /// Configures up to which size request bodies should be included in events.
+  /// This does not change wether an event is captured.
+  final MaxRequestBodySize maxRequestBodySize;
+
+  /// Describes which HTTP status codes should be considered as a failed
+  /// requests.
+  ///
+  /// Per default no status code is considered a failed request.
+  final List<SentryStatusCode> failedRequestStatusCodes;
+
+  /// Configures wether default PII is enabled for this client adapter.
+  final bool sendDefaultPii;
+
+  @override
+  Future<ResponseBody> fetch(
+    RequestOptions options,
+    Stream<Uint8List>? requestStream,
+    Future? cancelFuture,
+  ) async {
+    int? statusCode;
+    Object? exception;
+    StackTrace? stackTrace;
+
+    final stopwatch = Stopwatch();
+    stopwatch.start();
+
+    try {
+      final response =
+          await _client.fetch(options, requestStream, cancelFuture);
+      statusCode = response.statusCode;
+      return response;
+    } catch (e, st) {
+      exception = e;
+      stackTrace = st;
+      rethrow;
+    } finally {
+      stopwatch.stop();
+
+      // If captureFailedRequests is true, there statusCode is null.
+      // So just one of these blocks can be called.
+
+      if (captureFailedRequests && exception != null) {
+        await _captureEvent(
+          exception: exception,
+          stackTrace: stackTrace,
+          options: options,
+          requestDuration: stopwatch.elapsed,
+        );
+      } else if (failedRequestStatusCodes.containsStatusCode(statusCode)) {
+        final message =
+            'Event was captured because the request status code was $statusCode';
+        final httpException = SentryHttpClientError(message);
+
+        // Capture an exception if the status code is considered bad
+        await _captureEvent(
+          exception: exception ?? httpException,
+          options: options,
+          reason: message,
+          requestDuration: stopwatch.elapsed,
+        );
+      }
+    }
+  }
+
+  @override
+  void close({bool force = false}) => _client.close(force: force);
+
+  // See https://develop.sentry.dev/sdk/event-payloads/request/
+  Future<void> _captureEvent({
+    required Object? exception,
+    StackTrace? stackTrace,
+    String? reason,
+    required Duration requestDuration,
+    required RequestOptions options,
+  }) async {
+    // As far as I can tell there's no way to get the uri without the query part
+    // so we replace it with an empty string.
+    final urlWithoutQuery = options.uri.replace(query: '').toString();
+
+    final query = options.uri.query.isEmpty ? null : options.uri.query;
+
+    final headers = options.headers
+        .map((key, dynamic value) => MapEntry(key, value?.toString() ?? ''));
+
+    final sentryRequest = SentryRequest(
```

Comment:
Ideally we would record the content and content size but as far as I can tell, that's not possible. Is that right, @kuhnroyal ?